### PR TITLE
Chore: migrate tested bitcoin to new v0.21.1 tag

### PR DESCRIPTION
**Pull request by k9ert**
This PR will move the testing-backend for the cypress/cirrus/pytests to `v0.21.1`

### CODE REVIEW

**Comment by k9ert** [comment]
current status: After #1330 has been merged, this PR should be green. Next step: Remove conflicts.

**Comment by k9ert** [comment]
Current status: There is an interesting issue when sending a transaction:
```
>           txid = wallet.rpc.sendrawtransaction(txFS["hex"])
tests/test_specter.py:156: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
args = ('0200000001fd8eb4672d66be2cf40bd6691db0c38471e8ffbb4a042d1882046be5da78ca690000000000ffffffff02102700000000000017a91444c075d4228a40c5a46bf3e617c3597e240fee7e874bca052a010000001600143f1d10bc72e15851da48dcbb19fe4a6db9783fca00000000',)
kwargs = {}
r = {'error': {'code': -26, 'message': 'mandatory-script-verify-flag-failed (Operation not valid with the current stack size)'}, 'id': 0, 'result': None}
    def fn(*args, **kwargs):
        r = self.multi([(method, *args)], **kwargs)[0]
        if r["error"] is not None:
>           raise RpcError("Request error: %s" % r["error"]["message"], r)
E           cryptoadvance.specter.rpc.RpcError: Request error: mandatory-script-verify-flag-failed (Operation not valid with the current stack size)
src/cryptoadvance/specter/rpc.py:401: RpcError
------------------------------ Captured log call -------------------------------
```
@Sjors can you maybe help with that? What's wrong with this transaction?
Could not find time to decode it on my laptop (and don't know exactly whether it would make a difference, i don't think so) but i decoded with https://live.blockcypher.com/btc/decodetx/ like this:
```
{
    "addresses": [
        "37xYRxkiecPnUgisUfYHAuVw8BvPQqLMQv",
        "bc1q8uw3p0rju9v9rkjgmja3nlj2dkuhs072yf6k76"
    ],
    "block_height": -1,
    "block_index": -1,
    "confirmations": 0,
    "double_spend": false,
    "fees": 0,
    "hash": "bfc164e7bd4ce6a357262daf3fe24692ff7824dbb7114ab4b720d16a60cea980",
    "inputs": [
        {
            "age": 0,
            "output_index": 0,
            "prev_hash": "69ca78dae56b0482182d044abbffe87184c3b01d69d60bf42cbe662d67b48efd",
            "script_type": "empty",
            "sequence": 4294967295
        }
    ],
    "outputs": [
        {
            "addresses": [
                "37xYRxkiecPnUgisUfYHAuVw8BvPQqLMQv"
            ],
            "script": "a91444c075d4228a40c5a46bf3e617c3597e240fee7e87",
            "script_type": "pay-to-script-hash",
            "value": 10000
        },
        {
            "addresses": [
                "bc1q8uw3p0rju9v9rkjgmja3nlj2dkuhs072yf6k76"
            ],
            "script": "00143f1d10bc72e15851da48dcbb19fe4a6db9783fca",
            "script_type": "pay-to-witness-pubkey-hash",
            "value": 4999989835
        }
    ],
    "preference": "low",
    "received": "2021-09-06T10:21:11.956575341Z",
    "relayed_by": "54.234.186.128",
    "size": 114,
    "total": 4999999835,
    "ver": 2,
    "vin_sz": 1,
    "vout_sz": 2,
    "vsize": 114
}
```

**Comment by Sjors** [comment]
@k9ert not sure, but probably something wrong with the input. Maybe something went wrong with signing?

**Comment by k9ert** [comment]
> @k9ert not sure, but probably something wrong with the input. Maybe something went wrong with signing?

Thanks for your answer. I'm quite sure something in Bitcoin-Core changed. So the test went fine in v0.20.1 but fails in v0.21.1:
```
(.env) ➜  specter-desktop git:(master) ✗ ./tests/bitcoin/bin/bitcoin-cli -version              
Bitcoin Core RPC client version v0.21.1
(.env) ➜  specter-desktop git:(master) ✗ ./tests/install_noded.sh --bitcoin binary             
    --> install_noded.sh Start Mi 8. Sep 12:59:51 CEST 2021 (binary) for node_impl bitcoin
Checking for wget: install ok installed
    --> install version v0.20.1
    --> Listing binaries
-rwxr-xr-x 1 kim kim 10312344 Jul 31  2020 ./bitcoin/bin/bitcoind
-rwxr-xr-x 1 kim kim 5952832 Jul 31  2020 ./bitcoin/bin/bitcoin-wallet
-rwxr-xr-x 1 kim kim 16702880 Jul 31  2020 ./bitcoin/bin/test_bitcoin
-rwxr-xr-x 1 kim kim 31843176 Jul 31  2020 ./bitcoin/bin/bitcoin-qt
-rwxr-xr-x 1 kim kim 2376152 Jul 31  2020 ./bitcoin/bin/bitcoin-tx
-rwxr-xr-x 1 kim kim 2055800 Jul 31  2020 ./bitcoin/bin/bitcoin-cli
    --> checking for bitcoind
not found
    --> Finished installing bitcoind binary
    --> install_noded.sh End Mi 8. Sep 12:59:52 CEST 2021 took .581581256
(.env) ➜  specter-desktop git:(master) ✗ cat pytest.ini | grep bitcoind-version                                                                             
addopts = --bitcoind-version v0.20.1 --elementsd-version v0.20.99
(.env) ➜  specter-desktop git:(master) ✗ ./tests/bitcoin/bin/bitcoin-cli -version 
Bitcoin Core RPC client version v0.20.1
(.env) ➜  specter-desktop git:(master) ✗ pytest tests/test_specter.py -k test_abandon_purged_tx
======================================================================================= test session starts =======================================================================================
platform linux -- Python 3.8.10, pytest-6.1.2, py-1.10.0, pluggy-0.13.1
rootdir: /home/kim/src/specter-desktop, configfile: pytest.ini
plugins: cov-2.10.1
collected 3 items / 2 deselected / 1 selected                                                                                                                                                     

tests/test_specter.py .                                                                                                                                                                     [100%]

======================================================================================== warnings summary =========================================================================================
.env/lib/python3.8/site-packages/flask/json/__init__.py:31
  /home/kim/src/specter-desktop/.env/lib/python3.8/site-packages/flask/json/__init__.py:31: DeprecationWarning: Importing 'itsdangerous.json' is deprecated and will be removed in ItsDangerous 2.1. Use Python's 'json' module instead.
    _slash_escape = "\\/" not in _json.dumps("/")

.env/lib/python3.8/site-packages/flask/json/__init__.py:61
  /home/kim/src/specter-desktop/.env/lib/python3.8/site-packages/flask/json/__init__.py:61: DeprecationWarning: Importing 'itsdangerous.json' is deprecated and will be removed in ItsDangerous 2.1. Use Python's 'json' module instead.
    class JSONEncoder(_json.JSONEncoder):

.env/lib/python3.8/site-packages/flask/json/__init__.py:103
  /home/kim/src/specter-desktop/.env/lib/python3.8/site-packages/flask/json/__init__.py:103: DeprecationWarning: Importing 'itsdangerous.json' is deprecated and will be removed in ItsDangerous 2.1. Use Python's 'json' module instead.
    class JSONDecoder(_json.JSONDecoder):

.env/lib/python3.8/site-packages/pgpy/packet/fields.py:101
  /home/kim/src/specter-desktop/.env/lib/python3.8/site-packages/pgpy/packet/fields.py:101: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
    class SubPackets(collections.MutableMapping, Field):

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================================================================== 1 passed, 2 deselected, 4 warnings in 8.69s ===========================================================================
(.env) ➜  specter-desktop git:(master) ✗ vi pytest.ini
(.env) ➜  specter-desktop git:(master) ✗ cat pytest.ini | grep bitcoind-version                
addopts = --bitcoind-version v0.21.1 --elementsd-version v0.20.99
(.env) ➜  specter-desktop git:(master) ✗ ./tests/install_noded.sh --bitcoin binary             
    --> install_noded.sh Start Mi 8. Sep 13:02:39 CEST 2021 (binary) for node_impl bitcoin
Checking for wget: install ok installed
    --> install version v0.21.1
    --> Listing binaries
-rwxr-xr-x 1 kim kim 15277080 Apr 29 21:29 ./bitcoin/bin/bitcoind
-rwxr-xr-x 1 kim kim 10000192 Apr 29 21:29 ./bitcoin/bin/bitcoin-wallet
-rwxr-xr-x 1 kim kim 23208384 Apr 29 21:29 ./bitcoin/bin/test_bitcoin
-rwxr-xr-x 1 kim kim 37090632 Apr 29 21:29 ./bitcoin/bin/bitcoin-qt
-rwxr-xr-x 1 kim kim 4792088 Apr 29 21:29 ./bitcoin/bin/bitcoin-tx
-rwxr-xr-x 1 kim kim 4689528 Apr 29 21:29 ./bitcoin/bin/bitcoin-cli
    --> checking for bitcoind
not found
    --> Finished installing bitcoind binary
    --> install_noded.sh End Mi 8. Sep 13:02:39 CEST 2021 took .697376162
(.env) ➜  specter-desktop git:(master) ✗ ./tests/bitcoin/bin/bitcoin-cli -version              
Bitcoin Core RPC client version v0.21.1
(.env) ➜  specter-desktop git:(master) ✗ pytest tests/test_specter.py -k test_abandon_purged_tx
============================= test session starts ==============================
platform linux -- Python 3.8.10, pytest-6.1.2, py-1.10.0, pluggy-0.13.1
rootdir: /home/kim/src/specter-desktop, configfile: pytest.ini
plugins: cov-2.10.1
collected 3 items / 2 deselected / 1 selected

tests/test_specter.py F                                                  [100%]

=================================== FAILURES ===================================
________________________ test_abandon_purged_tx[False] _________________________
[...]
-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================== short test summary info ============================
FAILED tests/test_specter.py::test_abandon_purged_tx[False] - cryptoadvance.s...
================= 1 failed, 2 deselected, 4 warnings in 2.10s ==================
```

**Review by stepansnigirev — APPROVED** [review]
LGTM, looks like tests pass now

File: src/cryptoadvance/specter/util/common.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+def str2bool(my_str):
+    """returns a reasonable boolean from a string so that "False" will result in False"""
+    if my_str is None:
+        return False
+    elif isinstance(my_str, str) and my_str.lower() == "false":
+        return False
+    elif bool(my_str) is False:
+        return False
```
[the remark this answers is not in the document]
Comment (k9ert) [reply src/cryptoadvance/specter/util/common.py]: agree, resolved

**Comment by Sjors** [comment]
> LGTM, looks like tests pass now

Did you figure out what Bitcoin Core change was the issue?

File: src/cryptoadvance/specter/config.py [hunk 2]
```diff
@@ -148,6 +152,10 @@ class TestConfig(BaseConfig):
     # API active by default in test-mode
     SPECTER_API_ACTIVE = _get_bool_env_var("SPECTER_API_ACTIVE", "True")
 
+    # See #1316 since Bitcoin v0.21.1 (not only) the importmulti-call takes longer than 10 seconds on cirrus
```
Comment (Sjors) [inline src/cryptoadvance/specter/config.py:155]: If you can create a standalone example of this happening, please file a Github issue on https://github.com/bitcoin/bitcoin and tag me and @achow101 on it.

**Comment by stepansnigirev** [comment]
> > LGTM, looks like tests pass now
> 
> Did you figure out what Bitcoin Core change was the issue?

From my understanding sqlite wallets are behaving slightly differently - legacy wallets can generate new addresses of any type,
sqlite wallets can only generate wallets of the type of the imported descriptors, but `getnewaddress()` rpc call is using `bech32` as default address type, so if the test is using nested segwit wallets, `getnewaddress` rpc call will fail without arguments - you must provide address type `p2sh-segwit`.

Also there were issues with signing of raw transactions with the hot wallet if the hot wallet doesn't know about corresponding addresses - so I rewrote the test to convert raw tx to psbt, fill it with derivations and only then sign. We are not using rawtransactions in Specter itself, so doesn't make sense to use that in the tests as well.

**Comment by Sjors** [comment]
> `getnewaddress` rpc call will fail without arguments - you must provide address type `p2sh-segwit`

Ok, that behavior makes sense to me. As an RPC user, you should probably always specify `address_type` anyway. However, the default can be changed with `-addresstype`.

> if the hot wallet doesn't know about corresponding addresses

If you mean: "the hot wallet does not contain descriptors for these addresses" then the new behavior is correct. Legacy wallets eagerly sign anything they can, descriptor wallets are more picky.

> I rewrote the test to convert raw tx to psbt

That's probably more robust anyway.

**Comment by stepansnigirev** [comment]
> > `getnewaddress` rpc call will fail without arguments - you must provide address type `p2sh-segwit`
> 
> Ok, that behavior makes sense to me. As an RPC user, you should probably always specify `address_type` anyway. However, the default can be changed with `-addresstype`.
> 
> > if the hot wallet doesn't know about corresponding addresses
> 
> If you mean: "the hot wallet does not contain descriptors for these addresses" then the new behavior is correct. Legacy wallets eagerly sign anything they can, descriptor wallets are more picky.
> 
> > I rewrote the test to convert raw tx to psbt
> 
> That's probably more robust anyway.

Yes, no complaints about Bitcoin Core - everything makes sense. And also gave us an opportunity to improve tests :)

**Comment by k9ert** [comment]
>> getnewaddress rpc call will fail without arguments - you must provide address type p2sh-segwit

> Ok, that behavior makes sense to me. As an RPC user, you should probably always specify address_type anyway. However, the default can be changed with -addresstype.

If it's defacto a mandatory argument to get a reasonable result, why isn't it failing when it's not given? Wouldn't it be an improvement if it would fail with a reasonable error-message?

P.S.: I'm planning to create a test measuring the time difference for core before/after 0.21.1.

**Comment by Sjors** [comment]
> If it's defacto a mandatory argument to get a reasonable result, why isn't it failing when it's not given? Wouldn't it be an improvement if it would fail with a reasonable error-message?

By default a wallet has descriptors for each of the address types, so `getnewaddress` won't fail. But if you create a blank wallet and then import wallets, it will fail. A GitHub issue asking for a more clear error message makes sense though.

**Comment by k9ert** [comment]
After repackaging all the observability and ci_improvements to https://github.com/cryptoadvance/specter-desktop/pull/1403 , it became clear that the performance-issues did not came from Core v1.21.1 but were present for the older version as well and finally got resolved by increasing the resources on cirrus-ci (6CPU/6GB for the cypress-tests).
After merging that to this PR, this PR became trivial and only has changes in two files: `pytest.ini` and the test Stepan fixed.

@Sjors Thank you for your help, very much appreciated!

**Pull request merged**